### DISCUSSION TITLE
CallerArgumentExpression on generic type parameters

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#5298] [Language Ideas]
It would be really nice if the following was possible:
```
string Exp<T>([CallerArgumentExpression(nameof(T))] string? Expression = default) {
    return Expression;
}
```

Any chance someone could pick it up?

**Comment by CyrusNajmabadi** [comment]
What would you like this to do?

**Reply by TonyValenti** [reply]
Return the expression provided. The type name of T.

**Reply by CyrusNajmabadi** [reply]
I'm curious what the value is with that. Can you give a scenario where you need that or how you would use it?

**Reply by 333fred** [reply]
Can you tell me why `typeof(T).Name` doesn't suffice?

**Reply by TonyValenti** [reply]
I'm working on a project that uses IL weaving and obfuscation. The result is that type names get may led but we still have some places where we want to show the actual type name.  CallerArgumentExpression turns into a string which isn't mangled.

**Reply by svick** [reply]
@TonyValenti `typeof(T).Name` produces fairly simple IL. Couldn't your IL weaver turn that into a constant string before obfuscation?

**Reply by TonyValenti** [reply]
@svick - I'm not aware of a feature in PostSharp that can do that.

**Comment by Youssef1313** [comment]
I'm not the best to discuss language design issues, but I think this can be problematic:

- `CallerArgumentExpression` really returns the string representation as it appears in the syntax tree. But, what if the generic type is inferred? Would we disallow inference if the generic parameter is referenced in `CallerArgumentExpressionAttribute`? Would we produce empty string or null for this case? I don't think any of these options are good.
- If you're expecting a reliable value for the parameter, I don't think `CallerArgumentExpression` will satisfy your needs. The type can be specified via a using alias, in that case it won't return the name. The resulting string can change depending on whether the type is fully qualified or not.
- As it was pointed in other comments, `typeof(T).Name` can be more reliable and producing more consistent results.

**Comment by CyrusNajmabadi** [comment]
I think the core difference here is that expressions produce a value. The callee generally wants the value, but occasionally wants the expression that produced it. 

There is no such correspondence with type parameters. The caller and caller both are just referring to a type . And that type is passed on directly without evaluation. 

As such, there's no substantive difference between typeof(T), and any additional data that the compiler would pass along.

**Reply by dskinner** [reply]
> And that type is passed on directly without evaluation

```
T SomeMethod<T>() { /* typeof(T) has no access to field names below */ }
---
var x = SomeMethod<(int a, int b)>();
```

While the generic param is referred to as the "definition of a tuple type" in docs (I haven't looked at language spec) and not an "expression", it still stands that there is a loss of information from what was typed to what was received.

**Reply by siegfriedpammer** [reply]
In this case the tuple element names will not be preserved by the compiler anyway. Local variables cannot have attributes applied to them due to a metadata/runtime limitation, so the compiler must erase them.

See https://sharplab.io/#v2:EYLgtghglgdgNAFxFANnAJiA1AHwAIBMAjALABQeAzAASHUDC1A3udW9QCrUDKA9mAFMAsgIQALXugA8HAHwAKAJTNW7NXgDs1dAIBmEAK4oEAblVsAvuXPUbVWgBZqQpSrJqbANwgAnagA9qAF4efmFRCWl5WARqCDhqGOpgRQVFM3d2PCIATnl/ADoIdJsLWzILIA=

**Reply by dskinner** [reply]
> tuple element names will not be preserved by the compiler anyway

I'm under the impression from the CallerArgumentExpression proposal the compiler would emit such as follows if this did work, so what it preserves afterwards for runtime due to attribute limitations wouldn't be relevant (or is there a loss during multiple steps during compile?).

```
SomeMethod<(int, int)>(expression: "(int a, int b)")
```

This would be expanding the scope of the original proposal past expressions. Whether that's proper or not is another matter.

**Reply by ZacharyPatten** [reply]
I agree with Cyrus that currently I don't think this is worth it... _however_...

If we ever get anonymous methods in generic arguments like suggested here https://github.com/dotnet/csharplang/discussions/2904 then I think this topic would be necessary.
